server/src/main/java/com/google/fhir/gateway/AuditEventHelperImpl.java

```diff
@@ -0,0 +1,425 @@
+                .parseResource(transientDeleteResourceRaw);
+
+        if (transientDeleteResource instanceof DomainResource) {
+          boolean serverError =
+              !resources.isEmpty() && resources.get(0) instanceof OperationOutcome;
+          DomainResource resource =
+              serverError
+                  ? (DomainResource) resources.get(0)
+                  : (DomainResource) transientDeleteResource;
+
+          Set<String> patientIds = getPatientCompartmentOwners(resource);
+
+          if (!patientIds.isEmpty()) {
+            auditEventList.add(
+                createAuditEventCRUD(
+                    requestDetailsReader,
+                    resource,
+                    BalpProfileEnum.PATIENT_DELETE,
+                    agentUserWho,
+                    patientIds));
+          } else {
+            auditEventList.add(
+                createAuditEventCRUD(
+                    requestDetailsReader,
+                    resource,
+                    BalpProfileEnum.BASIC_DELETE,
+                    agentUserWho,
+                    null));
+          }
+        }
+
+        break;
+
+      default:
+        break;
+    }
+
+    // TODO Investigate bulk saving (batch processing) instead to improve performance. We'll
+    // probably need a mechanism for chunking e.g. 100, 200, or 500 batch
+    for (AuditEvent auditEvent : auditEventList) {
+      auditEvent.getPeriod().setEnd(new Date());
+      this.iGenericClient.create().resource(auditEvent).encodedJson().execute();
+    }
+  }
+
+  private AuditEvent.AuditEventOutcome mapOutcomeErrorCode(
+      OperationOutcome.IssueSeverity issueSeverity) {
+    AuditEvent.AuditEventOutcome errorCode = null;
+    if (OperationOutcome.IssueSeverity.FATAL.equals(issueSeverity)) {
+
+      errorCode = AuditEvent.AuditEventOutcome._12;
+    } else if (OperationOutcome.IssueSeverity.ERROR.equals(issueSeverity)) {
+
+      errorCode = AuditEvent.AuditEventOutcome._8;
+    } else if (OperationOutcome.IssueSeverity.WARNING.equals(issueSeverity)) {
+
+      errorCode = AuditEvent.AuditEventOutcome._4;
+    } else if (OperationOutcome.IssueSeverity.INFORMATION.equals(issueSeverity)) {
+
+      errorCode = AuditEvent.AuditEventOutcome._0;
+    }
+    return errorCode;
+  }
+
+  private List<IBaseResource> extractFhirResources(String serverContentResponse) {
+    List<IBaseResource> resourceList = new ArrayList<>();
+
+    IBaseResource responseResource =
+        this.iGenericClient.getFhirContext().newJsonParser().parseResource(serverContentResponse);
+
+    if (responseResource instanceof Bundle) {
+
+      resourceList =
+          ((Bundle) responseResource)
+              .getEntry().stream()
+                  .map(Bundle.BundleEntryComponent::getResource)
+                  .collect(Collectors.toList());
+
+    } else {
+      resourceList.add(responseResource);
+    }
+
+    return resourceList;
+  }
+
+  private Set<String> getCompartmentOwners(
```

Comment:
So it seems that you have the `patientFinder` instance variable but never use it; did you intend to swap this implementation with one that is `PatientFinder` based (I guess we should if we only need to deal with the patient compartment)?

Reply:
The current approach uses the FHIR native way to find the compartment owners from the resources. One thing we could do is move that logic as a helper method to complement what is in patientFinder and then invoke it from there. 

Also thought I might need patientFinder for scenarios the native approach doesn't handle easily e.g. bundle processing